app/src/test/java/com/quran/labs/androidquran/model/translation/ArabicDatabaseUtilsTest.kt

```diff
@@ -0,0 +1,123 @@
+
+import android.content.Context
+
+import com.quran.data.core.QuranInfo
+import com.quran.data.model.bookmark.Bookmark
+import com.quran.data.pageinfo.common.MadaniDataSource
+import com.quran.labs.androidquran.database.DatabaseHandler
+import com.quran.labs.androidquran.util.QuranFileUtils
+
+import org.junit.Before
+import org.junit.Test
+import org.mockito.Mock
+import org.mockito.Mockito.mock
+import org.mockito.MockitoAnnotations
+import com.google.common.truth.Truth.assertThat
+
+import java.util.ArrayList
+import java.util.HashMap
+
+class ArabicDatabaseUtilsTest {
+
+  @Mock
+  lateinit var context: Context
+
+  @Mock
+  lateinit var arabicHandler: DatabaseHandler
+
+  @Before
+  fun setUp() {
+    MockitoAnnotations.openMocks(this@ArabicDatabaseUtilsTest)
+  }
+
+  @Test
+  fun testHydrateAyahText() {
+    val arabicDatabaseUtils = getArabicDatabaseUtils()
+
+    val bookmarks = ArrayList<Bookmark>(3)
+    bookmarks.add(Bookmark(1, 1, 1, 1))
+    bookmarks.add(Bookmark(2, null, null, 3))
+    bookmarks.add(Bookmark(3, 114, 6, 604))
+
+    val result = arabicDatabaseUtils.hydrateAyahText(bookmarks)
+    assertThat(result).hasSize(3)
+
+    assertThat(result[0].ayahText).isNotEmpty()
+    assertThat(result[1].ayahText).isNull()
+
+    assertThat(result).isNotSameInstanceAs(bookmarks)
+  }
+
+  @Test
+  fun testHydrateAyahTextEmpty() {
+    val arabicDatabaseUtils = getArabicDatabaseUtils()
+
+    val bookmarks = ArrayList<Bookmark>(1)
+    bookmarks.add(Bookmark(1, null, null, 3))
+
+    val result = arabicDatabaseUtils.hydrateAyahText(bookmarks)
+    assertThat(result).hasSize(1)
+    assertThat(result[0].ayahText).isNull()
+    assertThat(result).isSameInstanceAs(bookmarks)
+  }
+
+  private fun getArabicDatabaseUtils(): ArabicDatabaseUtils {
+    return object : ArabicDatabaseUtils(context,
+      QuranInfo(MadaniDataSource()),
+      mock(QuranFileUtils::class.java)) {
+
+      override fun getArabicDatabaseHandler(): DatabaseHandler {
+        return arabicHandler
+      }
+
+      override fun getAyahTextForAyat(ayat: List<Int>): Map<Int, String> {
+        val result: MutableMap<Int, String> = HashMap()
+        for (ayahId in ayat) {
+          result[ayahId] = "verse $ayahId"
+        }
+        return result
+      }
+    }
+  }
+
+  @Test
+  fun testGetFirstFewWordsFromAyah() {
+    val total = ArabicDatabaseUtils.NUMBER_OF_WORDS
+    for (i in 1 until total) {
+      val text = makeText(i)
+      assertThat(ArabicDatabaseUtils.getFirstFewWordsFromAyah(4, 1, text)).isSameInstanceAs(text)
+    }
+
+    val veryLongString = makeText(100)
+    assertThat(ArabicDatabaseUtils.getFirstFewWordsFromAyah(4, 1, veryLongString))
+      .isEqualTo(makeText(4))
+  }
+
+  private fun makeText(words: Int): String {
+    val result = StringBuilder()
+
+    for (i in 0 until words) {
+      if (i > 0) {
+        result.append(" ")
+      }
+      result.append("word").append(i)
+    }
+
+    return result.toString()
```

Comment:
how about:

```kotlin
return words.map { "word$it" }.joinToString(" ")
```

Reply:
words is integer so it can not be mapped like that i think?

Reply:
sorry i meant `(0 until 5).map { ...`

Reply:
i see, commit has been pushed, thanks for suggesting that